File: go/pkg/genx/gemini.go
package genx

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"iter"
	"strings"

	"github.com/google/jsonschema-go/jsonschema"
	"github.com/googleapis/gax-go/v2/apierror"
	"google.golang.org/genai"
)

var _ Generator = (*GeminiGenerator)(nil)

// GeminiGenerator implements Generator using Google Gemini API.
type GeminiGenerator struct {
	Client *genai.Client `json:"-"`

	InvokeParams   *ModelParams `json:"invoke_params,omitzero"`
	GenerateParams *ModelParams `json:"generate_params,omitzero"`

	// Model should not start with "models/"
	Model string `json:"model"`
}

func (g *GeminiGenerator) Invoke(ctx context.Context, _ string, mctx ModelContext, fn *FuncTool) (Usage, *FuncCall, error) {
	cfg, contents, err := g.convModelContext(mctx)
	if err != nil {
		return Usage{}, nil, err
	}
	cfg.ResponseMIMEType = "application/json"
	cfg.ResponseSchema = geminiConvSchema(fn.Argument)
	resp, err := g.Client.Models.GenerateContent(ctx, g.Model, contents, cfg)
	if err != nil {
		if e, ok := err.(*apierror.APIError); ok {
			err = e.Unwrap()
		}
		return Usage{}, nil, err
	}
	if len(resp.Candidates) == 0 {
		return Usage{}, nil, fmt.Errorf("no candidates")
	}
	t := resp.Candidates[0]
	if t.FinishReason != genai.FinishReasonStop {
		if t.FinishReason == genai.FinishReasonMaxTokens {
			return geminiConvUsage(resp.UsageMetadata), nil, errors.New("max tokens")
		}
		return Usage{}, nil, fmt.Errorf("unexpected finish reason: %s", t.FinishReason)
	}
	var sb strings.Builder
	for _, p := range t.Content.Parts {
		if p.Text != "" {
			sb.WriteString(p.Text)
		}
	}
	return Usage{}, fn.NewFuncCall(sb.String()), nil
}

func (g *GeminiGenerator) GenerateStream(ctx context.Context, _ string, mctx ModelContext) (Stream, error) {
	cfg, contents, err := g.convModelContext(mctx)
	if err != nil {
		return nil, err
	}
	if len(contents) == 0 {
		return nil, fmt.Errorf("no contents")
	}
	sb := NewStreamBuilder(mctx, 32)
	go func() {
		if err := geminiPull(sb, g.Client.Models.GenerateContentStream(ctx, g.Model, contents, cfg)); err != nil {
			sb.Abort(err)
		}
	}()
	return sb.Stream(), nil
}

func geminiPull(builder *StreamBuilder, itr iter.Seq2[*genai.GenerateContentResponse, error]) error {
	var selIdx int32
	for chunk, err := range itr {
		if err != nil {
			return err
		}
		if len(chunk.Candidates) == 0 {
			continue
		}
		var sel *genai.Candidate
		if selIdx == 0 {
			selIdx = chunk.Candidates[0].Index
			sel = chunk.Candidates[0]
		} else {
			for _, c := range chunk.Candidates {
				if c.Index == selIdx {
					sel = c
					break
				}
			}
			if sel == nil {
				continue
			}
		}

		var (
			sb     strings.Builder
			blobs  = make(map[string]*bytes.Buffer)
			chunks = []*MessageChunk{}
		)
		for _, p := range sel.Content.Parts {
			switch {
			case p.Text != "":
				sb.WriteString(p.Text)
			case p.InlineData != nil:
				if _, ok := blobs[p.InlineData.MIMEType]; !ok {
					blobs[p.InlineData.MIMEType] = &bytes.Buffer{}
				}
				blobs[p.InlineData.MIMEType].Write(p.InlineData.Data)
			case p.FunctionCall != nil:
				b, _ := json.Marshal(p.FunctionCall.Args)
				chunks = append(chunks, &MessageChunk{
					Role: RoleModel,
					ToolCall: &ToolCall{
						ID: p.FunctionCall.Name,
						FuncCall: &FuncCall{
							Name:      p.FunctionCall.Name,
							Arguments: string(b),
						},
					},
				})
			default:
				return fmt.Errorf("unexpected part type: %T", p)
			}
		}
		if sb.Len() > 0 {
			chunks = append(chunks, &MessageChunk{
				Role: RoleModel,
				Part: Text(sb.String()),
			})
		}
		if len(blobs) > 0 {
			for mime, blob := range blobs {
				chunks = append(chunks, &MessageChunk{
					Role: RoleModel,
					Part: &Blob{
						MIMEType: mime,
						Data:     blob.Bytes(),
					},
				})
			}
		}
		if err := builder.Add(chunks...); err != nil {
			return err
		}
		switch sel.FinishReason {
		default:
			return builder.Unexpected(
				geminiConvUsage(chunk.UsageMetadata),
				fmt.Errorf("unexpected finish reason: %s", sel.FinishReason),
			)
		case genai.FinishReasonUnspecified, "":
			// continue
		case genai.FinishReasonStop:
			return builder.Done(geminiConvUsage(chunk.UsageMetadata))
		case genai.FinishReasonMaxTokens:
			return builder.Truncated(geminiConvUsage(chunk.UsageMetadata))
		case genai.FinishReasonSafety:
			var cats []string
			for _, sr := range sel.SafetyRatings {
				if sr.Blocked {
					cats = append(cats, string(sr.Category))
				}
			}
			return builder.Blocked(
				geminiConvUsage(chunk.UsageMetadata),
				"blocked by "+strings.Join(cats, ", "),
			)
		}
	}
	return errors.New("unexpected end of stream: no finish reason")
}

func geminiConvMessage(last *genai.Content, msg *Message) (new *genai.Content, err error) {
	var (
		role  string
		parts []*genai.Part
	)
	switch t := msg.Payload.(type) {
	default:
		return nil, fmt.Errorf("unexpected message type: %T", t)
	case Contents:
		switch msg.Role {
		default:
			return nil, fmt.Errorf("mismatched role and type: role=%s, type=%T", msg.Role, msg.Payload)
		case RoleUser:
			role = "user"
		case RoleModel:
			role = "model"
		}

		for _, c := range msg.Payload.(Contents) {
			switch v := c.(type) {
			case Text:
				parts = append(parts, genai.NewPartFromText(string(v)))
			case *Blob:
				parts = append(parts, genai.NewPartFromBytes(v.Data, v.MIMEType))
			}
		}
	case *ToolCall:
		role = "model"
		var args map[string]any
		if err := json.Unmarshal([]byte(t.FuncCall.Arguments), &args); err != nil {
			args = map[string]any{
				"text": t.FuncCall.Arguments,
			}
		}
		parts = append(parts, genai.NewPartFromFunctionCall(t.ID, args))
	case *ToolResult:
		role = "user"
		var result map[string]any
		if err := json.Unmarshal([]byte(t.Result), &result); err != nil {
			result = map[string]any{
				"text": t.Result,
			}
		}
		parts = append(parts, genai.NewPartFromFunctionResponse(t.ID, result))
	}
	if last == nil || last.Role != role {
		return &genai.Content{
			Role:  role,
			Parts: parts,
		}, nil
	}
	last.Parts = append(last.Parts, parts...)
	return nil, nil
}

func (g *GeminiGenerator) convModelContext(mctx ModelContext) (*genai.GenerateContentConfig, []*genai.Content, error) {
	cfg := genai.GenerateContentConfig{
		SafetySettings: []*genai.SafetySetting{
			{
				Category:  genai.HarmCategoryHateSpeech,
				Threshold: genai.HarmBlockThresholdOff,
			},
			{
				Category:  genai.HarmCategoryHarassment,
				Threshold: genai.HarmBlockThresholdOff,
			},
			{
				Category:  genai.HarmCategoryDangerousContent,
				Threshold: genai.HarmBlockThresholdOff,
			},
		},
	}
	prompts := []*genai.Part{}
	for p := range mctx.Prompts() {
		prompts = append(prompts, genai.NewPartFromText(p.Text))
	}
	if len(prompts) > 0 {
		cfg.SystemInstruction = &genai.Content{Parts: prompts}
	}
	mp := g.InvokeParams
	if p := mctx.Params(); p != nil {
		mp = p
	}
	if mp != nil {
		cfg.MaxOutputTokens = int32(mp.MaxTokens)
		cfg.Temperature = &mp.Temperature
		cfg.TopP = &mp.TopP
		cfg.TopK = &mp.TopK
	}

	tools := []*genai.Tool{}
	for t := range mctx.Tools() {
		switch t := t.(type) {
		case *FuncTool:
			tools = append(tools, &genai.Tool{
				FunctionDeclarations: []*genai.FunctionDeclaration{
					{
						Name:        t.Name,
						Description: t.Description,
						Parameters:  geminiConvSchema(t.Argument),
					},
				},
			})
		default:
			return nil, nil, fmt.Errorf("unexpected tool type: %T", t)
		}
	}
	if len(tools) > 0 {
		cfg.Tools = tools
	}

	var (
		contents []*genai.Content
		last     *genai.Content
	)
	for msg := range mctx.Messages() {
		new, err := geminiConvMessage(last, msg)
		if err != nil {
			return nil, nil, err
		}
		if new != nil {
			contents = append(contents, new)
			last = new
		}
	}
	if len(contents) == 0 {
		return nil, nil, fmt.Errorf("no contents")
	}

	return &cfg, contents, nil
}

func geminiConvSchema(schema *jsonschema.Schema) *genai.Schema {
	if schema == nil {
		return nil
	}

	enums := make([]string, 0, len(schema.Enum))
	for _, v := range schema.Enum {
		enums = append(enums, fmt.Sprintf("%v", v))
	}

	gs := genai.Schema{
		Format:      schema.Format,
		Description: schema.Description,
		Enum:        enums,
		Items:       geminiConvSchema(schema.Items),
		Required:    schema.Required,
	}

	requires := map[string]struct{}{}
	for _, v := range schema.Required {
		requires[v] = struct{}{}
	}

	if n := len(schema.Properties); n > 0 {
		gs.Properties = make(map[string]*genai.Schema, n)
		for k, prop := range schema.Properties {
			gs.Properties[k] = geminiConvSchema(prop)
		}
	}
	switch schema.Type {
	case "object":
		gs.Type = genai.TypeObject
	case "array":
		gs.Type = genai.TypeArray
	case "string":
		gs.Type = genai.TypeString
	case "number":
		gs.Type = genai.TypeNumber
	case "integer":
		gs.Type = genai.TypeInteger
	case "boolean":
		gs.Type = genai.TypeBoolean
	}
	return &gs
}

func geminiConvUsage(usage *genai.GenerateContentResponseUsageMetadata) Usage {
	return Usage{
		PromptTokenCount:        int64(usage.PromptTokenCount),
		CachedContentTokenCount: int64(usage.CachedContentTokenCount),
		GeneratedTokenCount:     int64(usage.CandidatesTokenCount),
	}
}

File: rust/genx/src/gemini.rs
//! Google Gemini Generator implementation.
//!
//! This module provides a generator for the Google Gemini API.
//!
//! # Example
//!
//! ```rust,ignore
//! use giztoy_genx::{Generator, ModelContextBuilder};
//! use giztoy_genx::gemini::{GeminiConfig, GeminiGenerator};
//!
//! let generator = GeminiGenerator::new(GeminiConfig {
//!     api_key: "AIza...".to_string(),
//!     model: "gemini-2.0-flash".to_string(),
//!     ..Default::default()
//! });
//!
//! let mut builder = ModelContextBuilder::new();
//! builder.prompt_text("system", "You are helpful.");
//! builder.user_text("user", "Hello!");
//!
//! let ctx = builder.build();
//! let mut stream = generator.generate_stream("", &ctx).await?;
//! ```

use async_trait::async_trait;
use futures::StreamExt;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::sync::Arc;

use crate::context::{ModelContext, ModelParams};
use crate::error::{GenxError, Usage};
use crate::stream::StreamBuilder;
use crate::tool::FuncTool;
use crate::types::{FuncCall, MessageChunk, Part, Payload, Role};
use crate::{Generator, Stream};

/// Gemini Generator configuration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GeminiConfig {
    /// API key for authentication
    pub api_key: String,
    /// Model name (e.g., "gemini-2.0-flash", "gemini-1.5-pro")
    pub model: String,
    /// Default generation parameters
    #[serde(default)]
    pub generate_params: Option<ModelParams>,
    /// Default invoke parameters
    #[serde(default)]
    pub invoke_params: Option<ModelParams>,
}

impl Default for GeminiConfig {
    fn default() -> Self {
        Self {
            api_key: String::new(),
            model: "gemini-2.0-flash".to_string(),
            generate_params: None,
            invoke_params: None,
        }
    }
}

/// Gemini Generator.
///
/// Supports the Google Gemini API for text generation.
pub struct GeminiGenerator {
    client: Client,
    config: GeminiConfig,
}

impl GeminiGenerator {
    /// Create a new Gemini generator.
    pub fn new(config: GeminiConfig) -> Self {
        Self {
            client: Client::new(),
            config,
        }
    }

    /// Get the configuration.
    pub fn config(&self) -> &GeminiConfig {
        &self.config
    }

    /// Convert model context to Gemini format.
    fn convert_request(&self, ctx: &dyn ModelContext, _stream: bool) -> Value {
        let mut contents = Vec::new();
        let mut system_instruction: Option<Value> = None;

        // Collect prompts as system instruction
        let prompt_texts: Vec<String> = ctx.prompts().map(|p| p.text.clone()).collect();
        if !prompt_texts.is_empty() {
            system_instruction = Some(json!({
                "parts": [{"text": prompt_texts.join("\n")}]
            }));
        }

        // Convert messages
        let mut current_role: Option<&str> = None;
        let mut current_parts: Vec<Value> = Vec::new();

        for msg in ctx.messages() {
            let role = match msg.role {
                Role::User => "user",
                Role::Model => "model",
                Role::Tool => "user", // Tool results come from user side in Gemini
            };

            // Flush previous content if role changes
            if current_role.is_some() && current_role != Some(role) {
                if !current_parts.is_empty() {
                    contents.push(json!({
                        "role": current_role.unwrap(),
                        "parts": current_parts,
                    }));
                    current_parts = Vec::new();
                }
            }
            current_role = Some(role);

            match &msg.payload {
                Payload::Contents(parts) => {
                    for part in parts {
                        match part {
                            Part::Text(t) => {
                                current_parts.push(json!({"text": t}));
                            }
                            Part::Blob(b) => {
                                current_parts.push(json!({
                                    "inline_data": {
                                        "mime_type": b.mime_type,
                                        "data": base64_encode(&b.data),
                                    }
                                }));
                            }
                        }
                    }
                }
                Payload::ToolCall(tc) => {
                    // Parse arguments as JSON
                    let args: Value =
                        serde_json::from_str(&tc.func_call.arguments).unwrap_or(json!({}));
                    current_parts.push(json!({
                        "functionCall": {
                            "name": tc.func_call.name,
                            "args": args,
                        }
                    }));
                }
                Payload::ToolResult(tr) => {
                    // Parse result as JSON
                    let result: Value = serde_json::from_str(&tr.result).unwrap_or(json!({
                        "text": tr.result
                    }));
                    current_parts.push(json!({
                        "functionResponse": {
                            "name": tr.id,
                            "response": result,
                        }
                    }));
                }
            }
        }

        // Flush remaining content
        if !current_parts.is_empty() {
            if let Some(role) = current_role {
                contents.push(json!({
                    "role": role,
                    "parts": current_parts,
                }));
            }
        }

        let params = ctx.params().or(self.config.generate_params.as_ref());

        let mut generation_config = json!({});
        if let Some(p) = params {
            if let Some(max) = p.max_tokens {
                generation_config["maxOutputTokens"] = json!(max);
            }
            if let Some(temp) = p.temperature {
                generation_config["temperature"] = json!(temp);
            }
            if let Some(top_p) = p.top_p {
                generation_config["topP"] = json!(top_p);
            }
            if let Some(top_k) = p.top_k {
                generation_config["topK"] = json!(top_k as i32);
            }
        }

        let mut body = json!({
            "contents": contents,
            "generationConfig": generation_config,
        });

        if let Some(sys) = system_instruction {
            body["systemInstruction"] = sys;
        }

        body
    }

    fn api_url(&self, stream: bool) -> String {
        let method = if stream {
            "streamGenerateContent"
        } else {
            "generateContent"
        };
        format!(
            "https://generativelanguage.googleapis.com/v1beta/models/{}:{}?key={}{}",
            self.config.model,
            method,
            self.config.api_key,
            if stream { "&alt=sse" } else { "" }
        )
    }
}

/// Simple base64 encoder.
fn base64_encode(data: &[u8]) -> String {
    const ALPHABET: &[u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

    let mut result = String::new();
    for chunk in data.chunks(3) {
        let mut n = (chunk[0] as u32) << 16;
        if chunk.len() > 1 {
            n |= (chunk[1] as u32) << 8;
        }
        if chunk.len() > 2 {
            n |= chunk[2] as u32;
        }

        result.push(ALPHABET[(n >> 18) as usize & 0x3F] as char);
        result.push(ALPHABET[(n >> 12) as usize & 0x3F] as char);

        if chunk.len() > 1 {
            result.push(ALPHABET[(n >> 6) as usize & 0x3F] as char);
        } else {
            result.push('=');
        }

        if chunk.len() > 2 {
            result.push(ALPHABET[n as usize & 0x3F] as char);
        } else {
            result.push('=');
        }
    }
    result
}

#[async_trait]
impl Generator for GeminiGenerator {
    async fn generate_stream(
        &self,
        _model: &str,
        ctx: &dyn ModelContext,
    ) -> Result<Box<dyn Stream>, GenxError> {
        let body = self.convert_request(ctx, true);
        let url = self.api_url(true);

        let response = self
            .client
            .post(&url)
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await
            .map_err(|e| GenxError::Other(anyhow::anyhow!("HTTP error: {}", e)))?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response
                .text()
                .await
                .unwrap_or_else(|_| "unknown".to_string());
            return Err(GenxError::Other(anyhow::anyhow!(
                "Gemini API error {}: {}",
                status,
                text
            )));
        }

        let builder = Arc::new(StreamBuilder::with_tools(32, vec![]));
        let builder_clone = builder.clone();

        // Spawn task to read SSE stream
        let mut stream = response.bytes_stream();
        tokio::spawn(async move {
            let mut buffer = String::new();

            while let Some(chunk_result) = stream.next().await {
                match chunk_result {
                    Ok(bytes) => {
                        buffer.push_str(&String::from_utf8_lossy(&bytes));

                        // Process lines (Gemini SSE format: "data: {...}\r\n" or "data: {...}\n")
                        while let Some(pos) = buffer.find('\n') {
                            let line = buffer[..pos].trim().to_string();
                            buffer = buffer[pos + 1..].to_string();

                            if line.is_empty() {
                                continue;
                            }

                            // Handle SSE data lines
                            let data = if let Some(d) = line.strip_prefix("data: ") {
                                d
                            } else if line.starts_with('{') {
                                // Sometimes Gemini sends raw JSON without "data: " prefix
                                &line
                            } else {
                                continue;
                            };

                            if let Ok(json) = serde_json::from_str::<Value>(data) {
                                // Extract text from candidates
                                if let Some(candidates) = json["candidates"].as_array() {
                                    for candidate in candidates {
                                        if let Some(parts) =
                                            candidate["content"]["parts"].as_array()
                                        {
                                            for part in parts {
                                                if let Some(text) = part["text"].as_str() {
                                                    let _ = builder_clone.add(&[
                                                        MessageChunk::text(Role::Model, text),
                                                    ]);
                                                }
                                            }
                                        }

                                        // Check finish reason
                                        if let Some(reason) = candidate["finishReason"].as_str() {
                                            match reason {
                                                "STOP" => {
                                                    let _ = builder_clone.done(Usage::default());
                                                    return;
                                                }
                                                "MAX_TOKENS" => {
                                                    let _ =
                                                        builder_clone.truncated(Usage::default());
                                                    return;
                                                }
                                                "SAFETY" => {
                                                    let _ = builder_clone.blocked(
                                                        Usage::default(),
                                                        "safety filter",
                                                    );
                                                    return;
                                                }
                                                _ => {}
                                            }
                                        }
                                    }
                                }

                                // Also check for error response
                                if json.get("error").is_some() {
                                    let err_msg = json["error"]["message"]
                                        .as_str()
                                        .unwrap_or("Unknown error");
                                    let _ = builder_clone.abort(GenxError::Other(anyhow::anyhow!(
                                        "Gemini error: {}",
                                        err_msg
                                    )));
                                    return;
                                }
                            }
                        }
                    }
                    Err(e) => {
                        let _ = builder_clone.abort(GenxError::Other(anyhow::anyhow!(
                            "Stream error: {}",
                            e
                        )));
                        return;
                    }
                }
            }

            let _ = builder_clone.done(Usage::default());
        });

        Ok(Box::new(builder.stream()))
    }

    async fn invoke(
        &self,
        _model: &str,
        ctx: &dyn ModelContext,
        tool: &FuncTool,
    ) -> Result<(Usage, FuncCall), GenxError> {
        let mut body = self.convert_request(ctx, false);

        // Add JSON schema for structured output
        body["generationConfig"]["responseMimeType"] = json!("application/json");
        body["generationConfig"]["responseSchema"] = tool.argument.clone();

        let url = self.api_url(false);

        let response = self
            .client
            .post(&url)
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await
            .map_err(|e| GenxError::Other(anyhow::anyhow!("HTTP error: {}", e)))?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response
                .text()
                .await
                .unwrap_or_else(|_| "unknown".to_string());
            return Err(GenxError::Other(anyhow::anyhow!(
                "Gemini API error {}: {}",
                status,
                text
            )));
        }

        let json: Value = response
            .json()
            .await
            .map_err(|e| GenxError::Other(anyhow::anyhow!("JSON parse error: {}", e)))?;

        // Extract text from response
        let content = json["candidates"][0]["content"]["parts"][0]["text"]
            .as_str()
            .ok_or_else(|| GenxError::Other(anyhow::anyhow!("No content in response")))?;

        Ok((Usage::default(), tool.new_func_call(content)))
    }
}
